### PR TITLE
Fix the number of completions generation by OpenAI API

### PR DESCRIPTION
**Pull request by nouhadziri**
This PR addresses two issues: it enables generating multiple completions from OpenAI and ensures that duplicate completions are skipped.

### CODE REVIEW

**Comment by vwxyzjn** [comment]
Do you want to rebase main? The files tab is showing a changes that have already happened.
image

**Comment by nouhadziri** [comment]
Would it be possible if you do it? I'm afraid to mess up things 🙈

**Comment by vwxyzjn** [comment]
Sure np let me take a look

**Comment by vwxyzjn** [comment]
I went ahead and fixed the merge conflicts and now doing some testing

File: open_instruct/rejection_sampling/api_generate.py [hunk 1]
```diff
@@ -6,13 +6,17 @@
...
+from open_instruct.rejection_sampling.prompt_templates import (
+    get_generation_template,
+    get_judgment_template,
+)
+
 
 @dataclass
 class LLMGenerationConfig:
-    n: int = 64
+    num_completions: int = 64
```
Comment (vwxyzjn) [inline open_instruct/rejection_sampling/api_generate.py:20]: I like the name changes! To be more consistent with the codebase, I changed it to `num_completions`

File: open_instruct/rejection_sampling/api_generate.py [hunk 2]
```diff
@@ -34,7 +38,7 @@ def __init__(self, config: LLMGenerationConfig):
         self.config = config
         self.async_client = AsyncOpenAI()
 
-    async def process_text(self, data: dict, i: int, limiter: asyncio.Semaphore, args: Args):
+    async def process_text(self, data: dict, i: int, limiter: asyncio.Semaphore, args: Args, gen_args: Args):
```
Comment (vwxyzjn) [inline open_instruct/rejection_sampling/api_generate.py:42]: Very nice addition.

**Review by vwxyzjn — CHANGES_REQUESTED** [review]
When running

```
python open_instruct/rejection_sampling.py \
    --input_filename output/completions.jsonl \
    --model_names_or_paths gpt-4 \
    --save_filename output/rejection_sampled_completions.jsonl \
    --num_completions 3 \
    --push_to_hub \
    --num_gpus 1
```

I got 

image



Could you investigate what went wrong? I noticed the reference completions are still judged multiple times. Could you point me to the logic to remove duplicate completions for reference completion judging?

image


When you run the RM script, it should give output like

processing: 0:64/216
processing: 64:128/216
processing: 128:192/216
processing: 192:256/216
processing: 0:64/72
processing: 64:128/72

Notice the 216 is the 3 completions per prompt, and the 72 is the reference completion for RM judges. Similarly, there should only be 72 requests to LLM as a judge for reference completions.

**Review by nouhadziri — COMMENTED** [review]
Thanks Costa for reviewing this and investigating the bug, could you approve this and I merge it? Thanks!

File: open_instruct/rejection_sampling/generation.py [hunk 3]
```diff
@@ -70,19 +75,27 @@ def print_rich_table(df: pd.DataFrame) -> Table:
...
-def generate_with_vllm(model_name_or_path: str, prompt_token_ids, gen_args: GenerationArgs):
+def generate_with_vllm(model_name_or_path: str, prompt_token_ids: List[int], gen_args: GenerationArgs):
     llm = LLM(model=model_name_or_path, tensor_parallel_size=gen_args.tensor_parallel_size)
+
+    # filter out prompts which are beyond the model's max token length
+    max_model_len = llm.llm_engine.scheduler_config.max_model_len
+    prompt_token_ids_len = len(prompt_token_ids)
+    prompt_token_ids = [item for item in prompt_token_ids if len(item) < max_model_len]
+    if len(prompt_token_ids) != prompt_token_ids_len:
+        print(f"Filtered out {prompt_token_ids_len - len(prompt_token_ids)} prompts which exceeds max token length")
```
Comment (vwxyzjn) [inline open_instruct/rejection_sampling/generation.py:93]: @jacob-morrison this is for you :D

**Review by vwxyzjn — APPROVED** [review]
Thanks @nouhadziri! I am doing some final testing and will merge after this.

File: open_instruct/rejection_sampling/generation.py (hunk 3 above)
Comment (nouhadziri) [reply open_instruct/rejection_sampling/generation.py:93]: Wonderful, thank you!!!

**Comment by vwxyzjn** [comment]
on jupiter https://beaker.org/ex/01J5C9BXR8MY9292734N4S84DA

on nfs https://beaker.org/ex/01J5C9BXR8MY9292734N4S84DA

all pass.

**Pull request merged**